Parser/pegen_errors.c

```diff
@@ -250,8 +250,15 @@ get_error_line_from_tokenizer_buffers(Parser *p, Py_ssize_t lineno)
     char *cur_line = p->tok->fp_interactive ? p->tok->interactive_src_start : p->tok->str;
     assert(cur_line != NULL);
 
-    for (int i = 0; i < lineno - 1; i++) {
-        cur_line = strchr(cur_line, '\n') + 1;
+    Py_ssize_t relative_lineno = p->starting_lineno ? lineno - p->starting_lineno + 1 : lineno;
+
+    for (int i = 0; i < relative_lineno - 1; i++) {
+        char *new_line = strchr(cur_line, '\n') + 1;
+        assert(new_line != NULL && new_line < p->tok->inp);
+        if (new_line == NULL || new_line >= p->tok->inp) {
```

Comment:
Aren't these two statements contrary to each other?

Reply:
Some context here: https://github.com/python/cpython/pull/30529#discussion_r782262467

Reply:
It would probably be worth adding a note then.  Otherwise I expect that most folks that read those two lines are going to be confused too.

Reply:
Yeah, one is asserting that everything goes ok (for debug builds), the other (for release builds) makes sure that if they are not ok, we don't crash.

Reply:
> It would probably be worth adding a note then. Otherwise I expect that most folks that read those two lines are going to be confused too.

Done in https://github.com/python/cpython/pull/30545